### PR TITLE
schema/schemaspec/schemahcl: patch references to blocks with variable names

### PR DESCRIPTION
**Pull request by rotemtam**
Fixes a bug in cases like:
```hcl
variable "tenant" {
	type = string
	default = "test"
}
schema "tenant" {
	name = var.tenant
}
table "users" {
	schema = schema.tenant
	column "id" {
		type = int
	}
}
```
When provided an input var such as `tenant=rotemtam`, the value of the reference to `schema.tenant` was `$schema.tenant` instead of `$schema.rotemtam`

### CODE REVIEW

File: schema/schemaspec/extension.go [hunk 1]
```diff
@@ -200,6 +200,30 @@ func (r *Resource) As(target interface{}) error {
 	return nil
 }
 
+// FinalName returns the final name for the resource by examining the struct tags for
+// the extension of the Resource's type. If no such extension is registered or the
+// extension struct does not have a name field, an error is returned.
+func (r *Resource) FinalName() (string, error) {
+	extensionsMu.RLock()
+	defer extensionsMu.RUnlock()
+	t, ok := extensions[r.Type]
```
Comment (a8m) [inline schema/schemaspec/extension.go:209]: Feels not right to me to synchronize codec flow. Don't you have this info on the Resource itself?

File: schema/schemaspec/schemahcl/hcl.go [hunk 2]
```diff
@@ -72,12 +72,71 @@ func (s *State) Eval(data []byte, v interface{}, input map[string]string) error
 	if err != nil {
 		return fmt.Errorf("schemahcl: failed decoding: %w", err)
 	}
+	if err := patchRefs(spec); err != nil {
+		return err
+	}
```
Comment (a8m) [inline schema/schemaspec/schemahcl/hcl.go:77]: I don't see any test case for this. Is it intentional?

File: schema/schemaspec/schemahcl/hcl.go [hunk 3]
```diff
@@ -72,12 +72,71 @@ func (s *State) Eval(data []byte, v interface{}, input map[string]string) error
...
 	if err := spec.As(v); err != nil {
 		return fmt.Errorf("schemahcl: failed reading spec as %T: %w", v, err)
 	}
 	return nil
 }
 
+// refMap maps addresses to their referenced resource.
+type refMap map[string]*schemaspec.Resource
```
Comment (a8m) [inline schema/schemaspec/schemahcl/hcl.go:85]: Don't add the type to the name, use something like `addrRef` instead.

File: schema/schemaspec/extension.go (hunk 1 above)
Comment (rotemtam) [reply schema/schemaspec/extension.go:209]: The lock protects concurrent reading and writing extensions (schemaspec.Register).
I did something similar to:
https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/database/sql/sql.go;l=817

File: schema/schemaspec/schemahcl/hcl.go (hunk 2 above)
Comment (rotemtam) [reply schema/schemaspec/schemahcl/hcl.go:77]: it's tested in the drivers, but I will add a specific case in the `schemahcl` package, you're right.

File: schema/schemaspec/extension.go (hunk 1 above)
Comment (rotemtam) [reply schema/schemaspec/extension.go:209]: discussed offlien, will remove this in a later pr

**Review by a8m — APPROVED** [review]

**Pull request merged**